crates/pop-cli/src/wallet_integration.rs

```diff
@@ -0,0 +1,553 @@
+use axum::{
+	http::HeaderValue,
+	response::Html,
+	routing::{get, post},
+	Router,
+};
+use serde::Serialize;
+use std::{path::PathBuf, sync::Arc};
+use tokio::{
+	sync::{oneshot, Mutex},
+	task::JoinHandle,
+};
+use tower_http::{cors::Any, services::ServeDir};
+
+/// Make frontend sourcing more flexible by allowing a custom route
+/// to be defined.
+pub trait Frontend {
+	fn serve_content(&self) -> Router;
+}
+
+/// Transaction payload to be sent to frontend for signing.
+#[derive(Serialize, Debug)]
+#[cfg_attr(test, derive(serde::Deserialize, Clone))]
+pub struct TransactionData {
+	chain_rpc: String,
+	call_data: Vec<u8>,
+}
+
+impl TransactionData {
+	pub fn new(chain_rpc: String, call_data: Vec<u8>) -> Self {
+		Self { chain_rpc, call_data }
+	}
+	#[allow(dead_code)]
+	pub fn call_data(&self) -> Vec<u8> {
+		self.call_data.clone()
+	}
+}
+
+/// Shared state between routes. Serves two purposes:
+/// - Maintains a channel to signal shutdown to the main app.
+/// - Stores the signed payload received from the wallet.
+#[derive(Default)]
+pub struct StateHandler {
+	/// Channel to signal shutdown to the main app.
+	shutdown_tx: Option<oneshot::Sender<()>>,
+	/// Received from UI.
+	pub signed_payload: Option<String>,
+	/// Holds a single error message.
+	/// Only method for consuming error removes (takes) it from state.
+	error: Option<String>,
+}
+
+/// Manages the wallet integration for secure signing of transactions.
+pub struct WalletIntegrationManager {
+	/// Shared state between routes.
+	pub state: Arc<Mutex<StateHandler>>,
+	/// Node rpc address.
+	pub rpc_url: String,
+	/// Web server task handle.
+	pub task_handle: JoinHandle<anyhow::Result<()>>,
+}
+
+impl WalletIntegrationManager {
+	/// Launches a server for hosting the wallet integration. Server launched in separate task.
+	/// Uses default address of 127.0.0.1:9090.
+	/// # Arguments
+	/// * `frontend`: A frontend with custom route to serve content.
+	/// * `payload`: Payload to be sent to the frontend for signing.
+	///
+	/// # Returns
+	/// A `WalletIntegrationManager` instance, with access to the state and task handle for the
+	/// server.
+	pub fn new<F: Frontend>(frontend: F, payload: TransactionData) -> Self {
```

Comment:
Let's have a port as one of the parameters! We can that way always make sure we are using an available port.

Reply:
https://github.com/r0gue-io/pop-cli/pull/371/commits/0e19b8a6cc693a545ffa9cd62eaff9e744127fb6